### PR TITLE
UCS/UCT/TEST: Refactor and optimize memory pool infrastructure.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/870/ for details.

**Comment by shamisp** [comment]
autobot build

**Comment by shamisp** [comment]
@yosefe - what is the main optimization here ?

**Comment by yosefe** [comment]
@shamisp see the commit message

**Comment by shamisp** [comment]
missed it. The description makes sense.

File: src/ucs/datastruct/mpool.c [hunk 1]
```diff
@@ -12,308 +13,242 @@
...
-              mp->name, mp->alignment, mp->max_elems, mp->elem_size, mp->elem_padding);
-    *mpp = mp;
+    ucs_debug("mpool %s: align %u, maxelems %u, elemsize %u",
+              ucs_mpool_name(mp), mp->data->alignment, max_elems, mp->data->elem_size);
     return UCS_OK;
 }
 
-static void __mpool_destroy(ucs_mpool_h mp)
+void ucs_mpool_cleanup(ucs_mpool_t *mp, int leak_check)
 {
```
Comment (shamisp) [inline src/ucs/datastruct/mpool.c]: #define UCS_MPOOL_LEAK_CHECK 1
#define UCS_MPOOL_NOLEAK_CHECK 0

**Comment by shamisp** [comment]
Besides the suggestion, everything else looks good.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/295/ for details. (Mellanox internal link)

File: src/ucs/datastruct/mpool.c (hunk 1 above)
Comment (yosefe) [inline src/ucs/datastruct/mpool.c]: i don't think it's needed because "leak_check" is a regular boolean value

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/299/ for details. (Mellanox internal link)

**Comment by shamisp** [comment]
:+1:

**Comment by shamisp** [comment]
obviously, you have to fix the merge issue first.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/885/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/308/ for details. (Mellanox internal link)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/307/ for details. (Mellanox internal link)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/886/ for details.

**Pull request merged**